R/updateNullModelApproxScoreSE.R

```diff
@@ -0,0 +1,246 @@
+## estimate the ratio (score.se.ratio) of the true SE(score) to the fast SE(score)
+## update the null model to contain this parameter
+## this needs to be run before using the approx.score.se option in assocTestSingle
+## this is based off of the SAIGE variance approximation
+
+updateNullModelApproxScoreSE <- function(gdsobj, null.model, nvar = 100, min.mac = 20, 
+     sparse=TRUE, imputed=FALSE, male.diploid=TRUE, genome.build=c("hg19", "hg38"), verbose=TRUE){
+
+     # check for W matrix
+     if (is.null(null.model$W)) stop('This null model was created with an older version of GENESIS and is not compatible with this analysis; 
+          please re-run your null model with the latest version.')
+
+     # samples in null model
+     sampid <- null.model$fit$sample.id
+     if(verbose) message(paste('null.model has', length(sampid), 'samples'))
+
+     # select a random set of variants meeting min.mac in the sample set
+     varid <- .selectRandomVars(gdsobj, sample.id = sampid, nvar = nvar, min.mac = min.mac, verbose = verbose)
+     if(verbose) message(paste('Selected', length(varid), 'variants with MAC >=', min.mac))
+
+     # calculate Score.SE using both approaches
+     tab <- .scoreSEtable(gdsobj, null.model, variant.id = varid, 
+                         sparse = sparse, imputed = imputed, male.diploid = male.diploid, genome.build = genome.build, verbose = verbose)
+	
+     # calculate the score.se.ratio estimate
+     score.se.ratio <- .scoreSEratio(tab)
+
+     # update the null model
+     null.model <- nullModelSmall(null.model)
+     null.model$score.se.ratio <- score.se.ratio
+     return(null.model)
+}
+
```

Comment:
I tried to write this function in a way that it could be used as is for a user with a small dataset on e.g. a laptop, but with sub-functions that would be amenable to creating a workflow on SBG.

On SBG, I'm imagining that we would do this analysis across chromosomes by creating 22 jobs that would each run `.selectRandomVars` and then `.scoreSEtable`. We would then have another job that would rbind those 22 output tables and pass that into `.scoreSEratio` and update the null.model. 

I went in circles for a while on the best way to implement this. We don't _have to_ make an updated null model, we could instead just compute that `score.se.ratio` parameter and pass it into `assocTestSingle` separately. However, one of the advantages of this approximation is that you can use a small null model. Perhaps updating the null model should be a separate function? I'm undecided what is best. One disadvantage of the way I have it now is that the function doesn't return the table of estimates for each variant (output of `.scoreSEtable`). I think that table is useful for the keen user, but maybe something a less savvy user would just ignore. 

Anyways, I'm very open to discussing the best way in which to put these steps together to make them usable.